### ISSUE TITLE
Possible to cache graphql.parse() results?

### ISSUE THREAD

**Issue by edulop91** [open-policy-agent/opa#5377]
## Short description 

I am wondering if there is an easy way to cache the graphql schema AST that is generated by `graphql.parse` function call. I have been working on integrating OPA for our graphql layer and, unfortunately, am running into performance issues due how long it takes to generate the AST from a raw query. 

Our graphQL schema does not change between each authorization decisions and therefore it makes sense to generate thes chema AST once and re-use it for each authorization query. 

I am running OPA embedded into Go mode. This is running locally from my laptop. 

## Steps To Reproduce

Nothing really interesting here. Rego looks like:
```rego
package graphql

import future.keywords

query_ast := graphql.parse(input.query, input.schema)[0]
```
The only difference is having farily large `input.query` and `input.schema`.

## Expected behavior

I would like to find a way to cache my graphql schema AST between authorization decisions.

**Comment by srenatus** [comment]
I think it should be possible to do this transparently. I.e. any call to `graphql.parse(..)` would check if the schema was already parsed, and retrieve that from the _inter-query cache_.

🤔 I wonder if this should be a feature to opt-in to, or if we need to worry about the cache growing too large... @philipaconrad What do you think?

**Comment by philipaconrad** [comment]
@srenatus I agree that we could probably borrow the inter-query cache system for this, but it easily could cause a memory usage spike for folks with many unique GraphQL schemas that get cached.

**Comment by srenatus** [comment]
Mentioned OOB -- maybe some sort of LRU cache could help here. If we cache 10 schemas, we'd probably solve the problem for most users, while not blowing up the memory usage for those special cases where many unique schemas are parsed. For those special cases, the situation would get neither better nor worse from having an LRU.

**Comment by adepretis** [comment]
I am hitting the same issue. A schema with about 2000 lines takes about 2-3 seconds to parse - for every request. Without some kind of cache this is not feasible :-(

**Comment by robmyersrobmyers** [comment]
In case it isn't obvious, you can pass the schema AST into OPA so it doesn't have to be parsed each time.  [Rego Playground Example](https://play.openpolicyagent.org/p/VkFUIw3IG3).

**Comment by anderseknert** [comment]
That'll still parse the schema every time though, won't it? Which won't be a problem in your case, but apparently so when the schema is thousands of lines.

If anyone would want to have a go at fixing this, I did something very similar some time ago for another built-in function. Could likely be mostly copy-pasted https://github.com/open-policy-agent/opa/pull/7081

**Comment by robmyersrobmyers** [comment]
Yep, but it helps a ton with the `running into performance issues due how long it takes to generate the AST from a raw query` aspect of the original issue.  It takes about 30s for OPA to generate the AST for the schema I'm testing with.  Thanks for the link to your previous work; I'll take a look.

**Comment by robmyersrobmyers** [comment]
I've got some initial results and wanted to see if I am on the right track before reworking the other functions in a similar fashion.

- Are there docs or best practices for putting things on the InterQueryValueCache?
  - For now I'm passing string inputs to `builtinCryptoSha256()` to compute a unique key
    - This results in the object representation and the string representation having different keys, but both can be cached
  - I could potentially use the text itself as a key, but it didn't seem like a great idea to put extra MBs in the cache
  - I am aware of the [pkg.go.dev](https://pkg.go.dev/github.com/open-policy-agent/opa/v1/topdown/cache) docs

```log
goarch: amd64
pkg: github.com/open-policy-agent/opa/v1/topdown
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkGraphQLSchemaIsValid/Trivial_Schema_-_string-16                     	   16966	     68579 ns/op
BenchmarkGraphQLSchemaIsValid/Trivial_Schema_with_cache_-_string-16         	 1000000	      1062 ns/op
BenchmarkGraphQLSchemaIsValid/Trivial_Schema_-_object-16                    	   12900	     92484 ns/op
BenchmarkGraphQLSchemaIsValid/Trivial_Schema_with_cache_-_object-16         	   92516	     12064 ns/op
BenchmarkGraphQLSchemaIsValid/Complex_Schema_-_string-16                    	      60	  19294987 ns/op
BenchmarkGraphQLSchemaIsValid/Complex_Schema_with_cache_-_string-16         	     342	   3443940 ns/op
PASS
ok  	github.com/open-policy-agent/opa/v1/topdown	12.613s
```
- For the complex benches I'm using the [GitHub GraphQL Schema](https://docs.github.com/en/graphql), which is about 1.2 M

**Comment by anderseknert** [comment]
Very nice @robmyersrobmyers 👍 No docs or best practices I'm aware of, but hashing the input seems like a good idea. The cache limit can be configured, but I'm not sure if it accounts for the key size (I'd have to check).. so keeping that small at the cost of hashing seems like a reasonable trade-off to me.

**Comment by robmyersrobmyers** [comment]
#### Status

I am still working on this.  I've got tests for all the graphql builtinss and benches for most of them, but I've still only added caching on one of the builtins.  

#### Questions

1. While testing with a complex schema I've run into an issue where it _appears_ that ast.InterfaceToValue() is called from builtinGraphQLParseSchema() and allocates a ton of memory and never returns.  I've made a [reproducer here](https://github.com/robmyersrobmyers/opa/blob/5377_cache_gql_parse/v1/topdown/main/reproducer.go).  Perhaps this representation of the schema causes a cycle in the tree somewhere?  Perhaps I'm doing something incorrectly?
2. For now, I'm [downloading and embedding](https://github.com/robmyersrobmyers/opa/blob/5377_cache_gql_parse/v1/topdown/graphql_bench_test.go#L24) large schemas for testing without adding them to the repo.  I figure we don't need to carry around a lot of extra MBs or incur any licensing or copyright issues with 3rd party GQL schemas.  Is there a preferred way to make it easy to benchmark with large schemas while not polluting the git repo?

**Comment by anderseknert** [comment]
Great work! 

Hmm, `ast.InterfaceToValue` never returning back doesn't sound good. There are no means in that to function to check for self-reference cycles, and I'm not sure how that would even be accomplished. But other than the map/slice handling, does it even deal with pointers? 🤔 I'd be very interested in knowing exactly where it gets stuck, and if we can reduce that to a minimal example. That never happens in the course of "normal" OPA handling of data.

As for the tests, I think the ideal soultion would be to generate the schema to use for benchmarking in the benchmark. I don't know much about GraphQL schemas though ... perhaps that's not feasible?

**Comment by anderseknert** [comment]
Excellent! Yeah if I wasn't clear before, I did in no way mean to imply that you should fix that self-referential infinite loop issue as part of your work :) That's a bug we'll need to look into separately. Looking forward to your PR — the numbers look incredible!

**Comment by robmyersrobmyers** [comment]
Status Update:

- I put in a separate [PR](https://github.com/open-policy-agent/opa/pull/7509) which I think resolves the ast.InterfaceToValue() issue.
- I still have some comments to work through on the PR that resolves this schema caching issue.

**Comment by StevenACoffman** [comment]
@robmyersrobmyers @johanfylling Just FYI, I merged into `github.com/vektah/gqlparser` the related PR:
* change: no longer include Position information when marshaling to JSON by @robmyersrobmyers in github.com/vektah/gqlparser#364

Which was released in 
https://github.com/vektah/gqlparser/releases/tag/v2.5.25

This was then updated in `github.com/99designs/gqlgen` in 99designs/gqlgen#3669
Which was released in [github.com/99designs/gqlgen v0.17.71](https://github.com/99designs/gqlgen/releases/tag/v0.17.71)
